[BE] 메뉴 API 서버 구현

yonghwankim-dev opened the pull request:
## 구현한것
- `/menus` API 구현

## 실행결과
![image](https://github.com/codesquad-gwanaksan/kiosk-max/assets/33227831/fc45fc52-5942-4e3b-aeb0-389910d72e76)

![image](https://github.com/codesquad-gwanaksan/kiosk-max/assets/33227831/95ed6451-59fc-47c1-9b21-a5d5540e732c)

yonghwankim-dev committed:
add: lombok 의존성 추가

- 게터, 생성자, 빌더 등을 편리하게 사용하기 위해서 추가하였습니다.

yonghwankim-dev committed:
feat: Product 객체 구현

- Product 객체는 키오스크 메뉴에 대한 정보를 가지고 있습니다.
- Category는 키오스크 메뉴에 대한 종류입니다.

yonghwankim-dev committed:
feat: ProductController 및 DTO 객체 구현

ProductController의 역할은 메뉴에 대한 정보를 응답하는 역할입니다.

yonghwankim-dev committed:
feat: ProductService, ProductRepository 구현

ProductService는 ProductRepository 인터페이스에 요청하여 메뉴에 대한 정보를 요청하는 역할입니다.

ProductRepository 역할은 DB 또는 메모리에 데이터를 저장하고 조회하는 역할입니다.

MemoryProductRepository 객체는 메모리 기반으로 Product 객체 정보를 관리하고 서비스 객체의 요청에 대해서 응답합니다.

yonghwankim-dev committed:
chore #5: mysql-connector-java 의존성 추가

yonghwankim-dev committed:
test #5: 상품 저장소 테스트 코드 추가

Jdbc 기반으로 상품이 저장되는지 테스트하는 코드를 추가하였습니다.

yonghwankim-dev committed:
test #5: 메모리 상품 저장소의 테스트 코드 수정

Category 클래스의 추가로 테스트 코드를 변경하였습니다.

yonghwankim-dev committed:
docs: applicatoin.properties 파일에 데이터 소스 설정 추가

yonghwankim-dev committed:
feat: Category 클래스 구현

Category 클래스는 메뉴별 종류의 정보를 가지는 클래스입니다.

yonghwankim-dev committed:
feat #5: Category 저장소 구현

구현체는 Jdbc 기반으로 작성하였습니다.

yonghwankim-dev committed:
feat #5: Jdbc 기반 상품 저장소 구현

yonghwankim-dev committed:
test #5: MenuDto로 클래스명 변경

yonghwankim-dev committed:
fix #5: 이미지 필드멤버명 imageUrl -> imaUrl로 변경

yonghwankim-dev committed:
feat #5: CategoryRepository findAll 구현

yonghwankim-dev committed:
fix #5: ProductDto -> MenuDto로 변경

yonghwankim-dev committed:
feat #5: MenuCategoryResponse 객체 구현

MenuCategoryResponse는 카테고리 이름, 카테고리 아이디, 카테고리에 따른 메뉴정보들을 저장하는 역할입니다.

yonghwankim-dev committed:
feat #5: MenuCategoryResponse 객체를 만들기 위한 컨트롤러 구현

yonghwankim-dev committed:
fix #5: applicatoin-local에서 데이터 소스 유저이름 및 비밀번호 변경

devbattery commented on be/src/main/java/com/kiosk/domain/entity/Product.java:
```diff
@@ -16,7 +16,7 @@ public class Product {
     private Long price;         // 가격
-    private String image;       // 이미지 저장 경로
+    private String imageUrl;       // 이미지 저장 경로
     private boolean isBest;     // 인기 상품
```
현재 json에서 `isBest`가 `best`로 넘어오는 상황입니다
네모님이랑 같이 해결해 볼게요!

yonghwankim-dev replied:
Jackson 라이브러리에서 직렬화 과정에서 getter/is와 같은 메소드를 이용하여 직렬화를 수행하는데 isBest와 같은 메소드에서 is를 없애고 best를 json 데이터의 제목으로 설정하여서 "isBest"가 아닌 "best"가 나온 것이었습니다. 

@JsonProperty 애노테이션을 사용하여 문제를 해결하였습니다.

yonghwankim-dev committed:
fix #5: Json에서 isBest -> best로 나오는 문제해결

문제원인은 Jackson 라이브러리에서 getter/is와 같은 메소드를 가지고 직렬화/역직렬화를 수행하는데 isBest와 같은 메소드에서 json 데이터를 만드는 과정에서 best로 제목을 정하기 때문입니다.

@JsonProperty 애노테이션을 사용하여 제목을 지정하여 문제를 해결했습니다.

yonghwankim-dev committed:
fix #5: applicatoin-local.properties 삭제

삭제 이유는 gitignore에 추가하기 위해서입니다.

yonghwankim-dev committed:
docs #5: gitignore에 applicatoin-local.properties 파일 추가

- 로컬 환경에서 데이터베이스를 사용하고 외부에서는 보이지 않게 하기 위해서 추가하였습니다.
- 대신 실행시 실행 환경에서 local 프로파일을 설정하여 실행할 수 있습니다.

yonghwankim-dev committed:
feat #5: KeyHolder를 이용한 데이터베이스에 데이터 삽입시 생성된 데이터의 pk 값을 반환하도록 구현

yonghwankim-dev committed:
refactor #5: 코드 정리

yonghwankim-dev committed:
test #5: Transactional 추가
